File: 1732-minimum-one-bit-operations-to-make-integers-zero/minimum-one-bit-operations-to-make-integers-zero.java
class Solution {
    public int minimumOneBitOperations(int n) {
        ArrayList<Long>f = new ArrayList<Long>(32);
        f.add(1L);
        for(int j = 1; j < 31; j++){
            f.add(f.get(j-1)*2 + 1);
        }

        int res=0;
        int plus=1;

        for(int j = 30; j>=0; j--){
            boolean isset = ((1<<j) & n) != 0;

            if(!isset){
                continue;
            }

            if(plus != 0){
                res+=f.get(j);
            }
            else{
                res-=f.get(j);
            }
            plus^=1;
        }

        return res;
    }
}

File: 1732-minimum-one-bit-operations-to-make-integers-zero/minimum-one-bit-operations-to-make-integers-zero.cpp
class Solution {
public:
    int minimumOneBitOperations(int n) {
        vector<long long>f(32);
        f[0]=1;
        for(int i=1; i<31; i++){
            f[i]=(f[i-1]*2+1);
        }
        
        int res = 0;
        int plus = 1;

        for(int i=30; i>=0; i--){
            bool isset = (1<<i & n) != 0;

            if(!isset){
                continue;
            }

            if(plus){
                res+=f[i];
            }
            else{
                res-=f[i];
            }
            plus^=1;
        }
        return res;
    }
};